src/overlays/actors/ovl_Obj_Chan/z_obj_chan.c

```diff
@@ -31,49 +39,404 @@ const ActorInit Obj_Chan_InitVars = {
+                                                          (this->actor.params & 0xFFF) | 0x1000, this->actor.cutscene,
+                                                          this->actor.unk20, &this->actor);
+        if (temp_v0 != NULL) {
+            this->pots[i] = temp_v0;
+            temp_v0->myPotIndex = i;
+            temp_v0->actor.cutscene = this->actor.cutscene;
+        } else {
+            Actor_MarkForDeath(&this->actor);
+        }
+    }
+
+    if (Flags_GetSwitch(globalCtx, this->actor.params & 0x7F)) {
+        this->stateFlags |= OBJCHAN_STATE_FIRE_DELAY;
+        this->stateFlags |= OBJCHAN_STATE_ON_FIRE;
+
+        this->rotationSpeed = OBJCHAN_ROTATION_SPEED;
+        this->flameSize = 1.0f;
+
+        for (i = 0; i < 5; i++) {
+            if (this->pots[i] != NULL) {
+                this->pots[i]->stateFlags |= OBJCHAN_STATE_FIRE_DELAY;
+                this->pots[i]->stateFlags |= OBJCHAN_STATE_ON_FIRE;
+                this->pots[i]->flameSize = 1.0f;
+            }
+        }
+    }
+    this->collider.dim.radius = 45;
+    this->collider.dim.height = 60;
+    this->collider.dim.yShift = -20;
+    this->actionFunc = ObjChan_MainAction;
+}
+
+void ObjChan_MainAction(ObjChan* thisx, GlobalContext* globalCtx) {
+    ObjChan* this = thisx;
+    ObjChan* temp;
+    s32 i;
+    Vec3f sp60;
+    Vec3s sp58;
+
+    if (this->unk1D0 > 0.0f) {
+        this->unk1D4 += 0x190;
+        if (this->unk1D0 <= 400.0f) {
+            this->unk1D0 = this->unk1D0 * 0.99f;
+        } else {
+            this->unk1D0 = this->unk1D0 - 1.0f;
+        }
+        if (this->unk1D0 < 0.0f) {
+            this->unk1D4 = 0;
+            this->unk1D0 = 0.0f;
+        }
+    }
+    this->actor.shape.rot.z = (Math_SinS(this->unk1D4) * this->unk1D0);
+    if ((this->stateFlags & OBJCHAN_STATE_START_CUTSCENE) && func_8013E2D4(&this->actor, this->cutscenes[0], -1, 0)) {
+        this->stateFlags |= OBJCHAN_STATE_CUTSCENE;
+        this->stateFlags &= ~OBJCHAN_STATE_START_CUTSCENE;
+    }
+    if ((this->stateFlags & OBJCHAN_STATE_CUTSCENE) && this->rotationSpeed == OBJCHAN_ROTATION_SPEED) {
+        this->stateFlags &= ~OBJCHAN_STATE_CUTSCENE;
+        ActorCutscene_Stop(this->cutscenes[0]);
+    }
+    Matrix_RotateY(this->actor.shape.rot.y, 0);
+    Matrix_InsertXRotation_s(this->actor.shape.rot.x, 1);
+    Matrix_InsertZRotation_s(this->actor.shape.rot.z, 1);
+    Matrix_RotateY(this->rotation, 1);
+    Matrix_GetStateTranslationAndScaledY(this->unk1CC, &this->actor.world.pos);
+    Math_Vec3f_Sum(&this->actor.world.pos, &this->unk1C0, &this->actor.world.pos);
+    Collider_UpdateCylinder(&this->actor, &this->collider);
+    for (i = 0; i < 5; i++) {
+        temp = this->pots[i];
+        if (temp != NULL) {
+            ObjChan_CalculatePotPosition(
+                &sp60, &sp58, &this->actor.world.pos, &this->actor.shape.rot,
+                (s16)((s32)((((f32)i * 360.0f) / 5.0f) * (65536.0f / 360.0f)) + this->rotation));
```

Comment:
```suggestion
                (s16)((s32)((((f32)i * 360.0f) / 5.0f) * (65536.0f / 360.0f)) + this->rotation));
```
Are all these casts necessary?